packages/engine/src/lib/variables/props-resolver.ts

```diff
@@ -0,0 +1,301 @@
+      ...resolveOptions,
+      variableName,
+    });
+    if (Array.isArray(resolvedValues[tokenPart])) {
+      longestResultLength = Math.max(
+        longestResultLength,
+        resolvedValues[tokenPart].length,
+      );
+    }
+  }
+  const result = new Array(longestResultLength).fill(null).map((_, index) => {
+    return Object.entries(resolvedValues).reduce((acc, [tokenPart, value]) => {
+      const valueToUse = (Array.isArray(value) ? value[index] : value) ?? '';
+      acc = acc.replace(
+        tokenPart,
+        isString(valueToUse) ? valueToUse : JSON.stringify(valueToUse),
+      );
+      return acc;
+    }, token);
+  });
+  return result;
+};
+
+export type PropsResolver = ReturnType<typeof createPropsResolver>;
+/**
+ * input: `Hello {{firstName}} {{lastName}}`
+ * tokenThatNeedResolving: [`{{firstName}}`, `{{lastName}}`]
+ */
+async function resolveInputAsync(
+  params: ResolveInputInternalParams,
+): Promise<unknown> {
+  const { input, currentState, engineToken, projectId, apiUrl, censoredInput } =
+    params;
+  const tokensThatNeedResolving = input.match(VARIABLE_PATTERN);
+  const inputContainsOnlyOneTokenToResolve =
+    tokensThatNeedResolving !== null &&
+    tokensThatNeedResolving.length === 1 &&
+    tokensThatNeedResolving[0] === input;
+  const resolveOptions = {
+    engineToken,
+    projectId,
+    apiUrl,
+    currentState,
+    censoredInput,
+  };
+
+  if (inputContainsOnlyOneTokenToResolve) {
+    const trimmedInput = input.trim();
+    const variableName = trimmedInput.substring(2, trimmedInput.length - 2);
+    return resolveSingleToken({
+      ...resolveOptions,
+      variableName,
+    });
+  }
+  const inputIncludesFlattenNestedKeysTokens = input.match(
+    FLATTEN_NESTED_KEYS_PATTERN,
+  );
+  if (
+    !isNil(inputIncludesFlattenNestedKeysTokens) &&
+    !isNil(tokensThatNeedResolving)
+  ) {
+    return mergeFlattenedKeysArraysIntoOneArray(
+      input,
+      tokensThatNeedResolving,
+      resolveOptions,
+    );
+  }
+
+  return replaceAsync(
+    input,
+    VARIABLE_PATTERN,
+    async (_fullMatch, variableName) => {
+      const result = await resolveSingleToken({
+        ...resolveOptions,
+        variableName,
+      });
+      return isString(result) ? result : JSON.stringify(result);
+    },
+  );
+}
+
+async function resolveSingleToken(
+  params: ResolveSingleTokenParams,
+): Promise<unknown> {
+  const { variableName, currentState } = params;
+  const isConnection = variableName.startsWith(CONNECTIONS);
+  if (isConnection) {
+    return handleConnection(params);
+  }
+  return evalInScope(variableName, { ...currentState });
+}
+
+async function handleConnection(
+  params: ResolveSingleTokenParams,
+): Promise<unknown> {
+  const { variableName, engineToken, projectId, apiUrl, censoredInput } =
+    params;
+  const connectionName = parseConnectionNameOnly(variableName);
+  if (isNil(connectionName)) {
+    return '';
+  }
+  if (censoredInput) {
```

Comment:
move this if above the const connectionName = parseConnectionNameOnly(variableName);

no need to do any addtional logic in that case